connman/plugins/sailfish_ofono.c

```diff
@@ -582,13 +599,102 @@ static int modem_configure(struct modem_data *md)
 	return index;
 }
 
+static gboolean is_modem_network_configured(struct modem_data *md)
+{
+	DBG("%p", md);
+
+	switch (md->connctx->protocol) {
+	case OFONO_CONNCTX_PROTOCOL_UNKNOWN:
+	case OFONO_CONNCTX_PROTOCOL_NONE:
+		return FALSE;
+	case OFONO_CONNCTX_PROTOCOL_IP:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV4)) {
+			connman_warn("ofono: %p IPv4 no address set", md);
+			return FALSE;
+		}
+
+		break;
+	case OFONO_CONNCTX_PROTOCOL_IPV6:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV6)) {
+			connman_warn("ofono: %p IPv6 no address set", md);
+			return FALSE;
+		}
+
+		break;
+	case OFONO_CONNCTX_PROTOCOL_DUAL:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV4)) {
+			connman_warn("ofono: %p DUAL no IPv4 address set", md);
+			return FALSE;
+		}
+
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV6)) {
+			connman_warn("ofono: %p DUAL no IPv6 address set", md);
+			return FALSE;
+		}
+
+		break;
+	}
+
+	DBG("%p is configured", md);
+
+	return TRUE;
+}
+
+static gboolean delayed_connect(gpointer data) {
+	struct modem_data *md = data;
+
+	/* Keep in loop until configured and attempt limit is not reached. */
+	if (!is_modem_network_configured(md) &&
+			md->delayed_connect_attempts < DELAYED_CONNECT_LIMIT) {
+		md->delayed_connect_attempts++;
+		return TRUE;
+	}
+
+	/* Log that we're giving up, network setup wasn't completed in time. */
+	if (md->delayed_connect_attempts == DELAYED_CONNECT_LIMIT)
+		connman_error("cellular setup was not completed in time");
+
+	DBG("modem %p network %p configured, set connected", md, md->network);
+
+	connman_network_set_connected(md->network, TRUE);
+
+	g_source_remove(md->delayed_connect_id);
+	md->delayed_connect_id = 0;
+	md->delayed_connect_attempts = 0;
+
+	return FALSE;
+}
+
+static void modem_set_connected(struct modem_data *md)
+{
+	if (!is_modem_network_configured(md)) {
+		if (md->delayed_connect_id)
+			g_source_remove(md->delayed_connect_id);
+
+		DBG("%p init delayed connect, modem network not ready", md);
+
+		md->delayed_connect_attempts = 0;
+		md->delayed_connect_id = g_timeout_add(DELAYED_CONNECT_TIMEOUT,
+							delayed_connect, md);
+	} else {
+		connman_network_set_connected(md->network, TRUE);
```

Comment:
Cancel the pending delayed config when it's no longer needed

Reply:
Added a function for this as it is to be done in similar fashion when disconnecting as well. When this is called we should anyways clean those values to either connect or start the delayed connect.